fearless/Modules/NewWallet/SendFlow/ChooseRecipient/ChooseRecipientViewController.swift

```diff
@@ -0,0 +1,196 @@
+    }
+
+    @objc private func pasteButtonClicked() {
+        rootView.searchView.textField.text = UIPasteboard.general.string
+    }
+
+    @objc private func historyButtonClicked() {
+        presenter.didTapHistoryButton()
+    }
+
+    @objc private func nextButtonClicked() {
+        guard let address = rootView.searchView.textField.text else {
+            return
+        }
+        presenter.didTapNextButton(with: address)
+    }
+}
+
+extension ChooseRecipientViewController: ChooseRecipientViewProtocol {
+    func didReceive(viewModel: ChooseRecipientViewModel) {
+        rootView.bind(viewModel: viewModel)
+    }
+
+    func didReceive(tableViewModel: ChooseRecipientTableViewModel) {
+        self.tableViewModel = tableViewModel
+        rootView.tableView.isHidden = tableViewModel.results.isEmpty
+        rootView.tableView.reloadData()
+    }
+
+    func didReceive(locale: Locale) {
+        self.locale = locale
+        rootView.locale = locale
+    }
+
+    func didReceive(address: String) {
+        rootView.searchView.textField.text = address
+    }
+}
+
+extension ChooseRecipientViewController: UITableViewDelegate, UITableViewDataSource {
+    func numberOfSections(in _: UITableView) -> Int {
+        guard let viewModel = tableViewModel, !viewModel.results.isEmpty else {
+            return 0
+        }
+        return 1
+    }
+
+    func tableView(_: UITableView, numberOfRowsInSection _: Int) -> Int {
+        guard let viewModel = tableViewModel, !viewModel.results.isEmpty else {
+            return 0
+        }
+        return viewModel.results.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let viewModel = tableViewModel else {
+            return UITableViewCell()
+        }
+        guard let cell = tableView.dequeueReusableCellWithType(SearchPeopleTableCell.self) else {
+            return UITableViewCell()
+        }
+        cell.bind(to: viewModel.results[indexPath.row])
+        return cell
+    }
+
+    func tableView(_: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let viewModel = tableViewModel else {
+            return
+        }
+        rootView.searchView.textField.text = viewModel.results[indexPath.row].address
+        presenter.didSelectViewModel(cellViewModel: viewModel.results[indexPath.row])
+    }
+}
+
+extension ChooseRecipientViewController: UITextFieldDelegate {
+    func textField(
+        _ textField: UITextField,
+        shouldChangeCharactersIn range: NSRange,
+        replacementString string: String
+    ) -> Bool {
+        guard let text = textField.text as NSString? else {
+            return true
+        }
+
+        let newString = text.replacingCharacters(in: range, with: string)
+
+        presenter.searchTextDidChanged(newString)
+        return true
+    }
+
+    func textFieldShouldClear(_: UITextField) -> Bool {
+        presenter.searchTextDidChanged("")
```

Comment:
use nil

Reply:
Where are not optional string, it's required for search protocol